### PR TITLE
Add discussion of ConfigureAwait

### LINKED ISSUE #13874: Does ConfigureAwait(false) make sense for async streams?
The examples with `await foreach` do not show if `ConfigureAwait` can be used.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: a94537c5-754e-9dea-94a8-7573011f588c
* Version Independent ID: aba09118-22c5-2c6a-954a-b9659a9644ae
* Content: [What's New in C# 8.0 - C# Guide](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-8#static-local-functions)
* Content Source: [docs/csharp/whats-new/csharp-8.md](https://github.com/dotnet/docs/blob/master/docs/csharp/whats-new/csharp-8.md)
* Product: **dotnet-csharp**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

### PR DESCRIPTION
**Pull request by BillWagner**
The ConfigureAwait method can be used in async streams.

Fixes #13874

### CODE REVIEW

**Review by tdykstra — APPROVED** [review]
LGTM.  One suggestion for you to consider.

File: docs/csharp/language-reference/keywords/foreach-in.md [hunk 1]
```diff
@@ -19,7 +19,7 @@ The `foreach` statement executes a statement or a block of statements for each e
 
 Beginning with C# 7.3, if the enumerator's `Current` property returns a [reference return value](ref.md#reference-return-values) (`ref T` where `T` is the type of the collection element), you can declare the iteration variable with the `ref` or `ref readonly` modifier.
 
-Beginning with C# 8.0, the `await` operator may be applied to the `foreach` statement when the collection type implements the <xref:System.Collections.Generic.IAsyncEnumerable%601> interface. Each iteration of the loop may be suspended while the next element is retrieved asynchronously. 
+Beginning with C# 8.0, the `await` operator may be applied to the `foreach` statement when the collection type implements the <xref:System.Collections.Generic.IAsyncEnumerable%601> interface. Each iteration of the loop may be suspended while the next element is retrieved asynchronously. The extension method <xref:System.Threading.Tasks.TaskAsyncEnumerableExtensions.ConfigureAwait%2A?displayProperty=nameWithType> enables you to control if consuming the stream executes on the captured context.
```
[the remark this answers is not in the document]
Comment (pkulikov) [reply docs/csharp/language-reference/keywords/foreach-in.md]: @BillWagner @tdykstra should we move the focus from `ConfigureAwait` to context capturing? Something like that:

> By default, stream elements are processed in the captured context. If you want to disable capturing of the context, use the <xref:System.Threading.Tasks.TaskAsyncEnumerableExtensions.ConfigureAwait%2A?displayProperty=nameWithType> extension method.

Two questions to the above suggestion:
- Is the described default behavior correct?
- Is it possible to make a xref link with the specified argument value, something like:
  `<xref:System.Threading.Tasks.TaskAsyncEnumerableExtensions.ConfigureAwait(false)?displayProperty=nameWithType>` 
  to make it render as `TaskAsyncEnumerableExtensions.ConfigureAwait(false)`?

[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/language-reference/keywords/foreach-in.md]: > Is it possible to make a xref link with the specified argument value, something like:
<xref:System.Threading.Tasks.TaskAsyncEnumerableExtensions.ConfigureAwait(false)?displayProperty=nameWithType>
to make it render as TaskAsyncEnumerableExtensions.ConfigureAwait(false)?

No, that's not supported. (Great idea though)

File: docs/csharp/language-reference/keywords/foreach-in.md [hunk 2]
```diff
@@ -19,7 +19,7 @@ The `foreach` statement executes a statement or a block of statements for each e
...
 
-Beginning with C# 8.0, the `await` operator may be applied to the `foreach` statement when the collection type implements the <xref:System.Collections.Generic.IAsyncEnumerable%601> interface. Each iteration of the loop may be suspended while the next element is retrieved asynchronously. 
+Beginning with C# 8.0, the `await` operator may be applied to the `foreach` statement when the collection type implements the <xref:System.Collections.Generic.IAsyncEnumerable%601> interface. Each iteration of the loop may be suspended while the next element is retrieved asynchronously. By default, stream elements are processed in the captured context. If you want to disable capturing of the context, use the xref:System.Threading.Tasks.TaskAsyncEnumerableExtensions.ConfigureAwait%2A?displayProperty=nameWithType extension method call. For more information on synchronization contexts and capturing the current context, see the article on [consuming the Task based asynchronous pattern](../../../standard/asynchronous-programming-patterns/consuming-the-task-based-asynchronous-pattern.md).
```
Comment (pkulikov) [inline docs/csharp/language-reference/keywords/foreach-in.md]:
```suggestion
Beginning with C# 8.0, the `await` operator may be applied to the `foreach` statement when the collection type implements the <xref:System.Collections.Generic.IAsyncEnumerable%601> interface. Each iteration of the loop may be suspended while the next element is retrieved asynchronously. By default, stream elements are processed in the captured context. If you want to disable capturing of the context, use the <xref:System.Threading.Tasks.TaskAsyncEnumerableExtensions.ConfigureAwait%2A?displayProperty=nameWithType> extension method. For more information about synchronization contexts and capturing the current context, see the article on [consuming the Task-based asynchronous pattern](../../../standard/asynchronous-programming-patterns/consuming-the-task-based-asynchronous-pattern.md).
```

File: docs/csharp/tutorials/generate-consume-asynchronous-stream.md [hunk 3]
```diff
@@ -126,6 +126,8 @@ Replace that code with the following `await foreach` loop:
 
 [!code-csharp[FinishedEnumerateAsyncStream](~/samples/snippets/csharp/tutorials/AsyncStreams/finished/IssuePRreport/IssuePRreport/Program.cs#EnumerateAsyncStream)]
 
+By default, stream elements are processed in the captured context. If you want to disable capturing of the context, use the xref:System.Threading.Tasks.TaskAsyncEnumerableExtensions.ConfigureAwait%2A?displayProperty=nameWithType extension method call. For more information on synchronization contexts and capturing the current context, see the article on [consuming the Task based asynchronous pattern](../../standard/asynchronous-programming-patterns/consuming-the-task-based-asynchronous-pattern.md).
```
Comment (pkulikov) [inline docs/csharp/tutorials/generate-consume-asynchronous-stream.md]:
```suggestion
By default, stream elements are processed in the captured context. If you want to disable capturing of the context, use the <xref:System.Threading.Tasks.TaskAsyncEnumerableExtensions.ConfigureAwait%2A?displayProperty=nameWithType> extension method. For more information about synchronization contexts and capturing the current context, see the article on [consuming the Task-based asynchronous pattern](../../standard/asynchronous-programming-patterns/consuming-the-task-based-asynchronous-pattern.md).
```

File: docs/csharp/whats-new/csharp-8.md [hunk 4]
```diff
@@ -385,7 +385,7 @@ await foreach (var number in GenerateSequence())
 }
 ```
 
-You can try asynchronous streams yourself in our tutorial on [creating and consuming async streams](../tutorials/generate-consume-asynchronous-stream.md).
+You can try asynchronous streams yourself in our tutorial on [creating and consuming async streams](../tutorials/generate-consume-asynchronous-stream.md). By default, stream elements are processed in the captured context. If you want to disable capturing of the context, use the xref:System.Threading.Tasks.TaskAsyncEnumerableExtensions.ConfigureAwait%2A?displayProperty=nameWithType extension method call. For more information on synchronization contexts and capturing the current context, see the article on [consuming the Task based asynchronous pattern](../../standard/asynchronous-programming-patterns/consuming-the-task-based-asynchronous-pattern.md).
```
Comment (pkulikov) [inline docs/csharp/whats-new/csharp-8.md]:
```suggestion
You can try asynchronous streams yourself in our tutorial on [creating and consuming async streams](../tutorials/generate-consume-asynchronous-stream.md). By default, stream elements are processed in the captured context. If you want to disable capturing of the context, use the <xref:System.Threading.Tasks.TaskAsyncEnumerableExtensions.ConfigureAwait%2A?displayProperty=nameWithType> extension method. For more information about synchronization contexts and capturing the current context, see the article on [consuming the Task-based asynchronous pattern](../../standard/asynchronous-programming-patterns/consuming-the-task-based-asynchronous-pattern.md).
```

**Review by pkulikov — COMMENTED** [review]
Fixed xref links and minor style updates.

**Pull request merged**